### PR TITLE
FYST-1041 ID form 40 add some lines

### PR DESCRIPTION
**Pull request by jenny-heath**
## Link to pivotal/JIRA issue
- https://codeforamerica.atlassian.net/browse/FYST-1041
## Is PM acceptance required? (delete one)
- Yes - don't merge until JIRA issue is accepted!

**Reminder**: merge main into this branch and get green tests before merging to main
## What was done?
- Adds lines 12a-c, 15-17, 19-21 to form ID40
- Some lines are only present on the PDF but not the XML
- Some lines are pulled directly from DF so don't have calculator methods
## How to test?
- idk if there's much that's not self-explanatory here

### CODE REVIEW

File: app/lib/efile/id/id40_calculator.rb [hunk 1]
```diff
@@ -85,6 +87,24 @@ def calculate_line_11
...
+        [line_or_zero(:ID40_LINE_11) - @direct_file_data.total_itemized_or_standard_deduction_amount, 0].max
+      end
+
+      WKSHT_LINE_2_AMTS = {
+        single: 4673,
+        married_filing_separately: 4673,
+        married_filing_jointly: 9346,
+        head_of_household: 9346,
+        qualifying_widow: 9346,
+      }.freeze
```
Comment (jenny-heath) [inline app/lib/efile/id/id40_calculator.rb:102]: this looks kind of stupid but it's much nicer to access? idk

File: app/models/direct_file_data.rb [hunk 2]
```diff
@@ -121,6 +123,14 @@ def phone_number=(value)
...
   end
 
+  def cell_phone_number=(value)
+    write_df_xml_value(__method__, value)
+  end
+
+  def tax_payer_email=(value)
+    write_df_xml_value(__method__, value)
+  end
+
```
Comment (jenny-heath) [inline app/models/direct_file_data.rb:133]: added these so i could test that the list of attrs in the test also have setters (see direct_file_data_spec)

File: spec/lib/submission_builder/ty2024/states/id/documents/id40_spec.rb [hunk 3]
```diff
@@ -107,29 +170,12 @@
     end
 
     context "PermanentBuildingFund" do
-      context "when a client is not blind has filing requirements and does not receive public assistance" do
```
Comment (jenny-heath) [inline spec/lib/submission_builder/ty2024/states/id/documents/id40_spec.rb:110]: we didn't need to test this again, it's already tested in the calculator spec

File: spec/models/direct_file_data_spec.rb [hunk 4]
```diff
@@ -52,6 +57,36 @@
...
+
+  describe "#is_spouse_blind?" do
+    it "returns true when value is X" do
+      expect(direct_file_data.is_spouse_blind?).to eq true
+    end
+
+    it "returns false when node absent" do
+      xml.at('SpouseBlindInd').remove
+      expect(direct_file_data.is_spouse_blind?).to eq false
+    end
```
Comment (jenny-heath) [inline spec/models/direct_file_data_spec.rb:89]: these methods had been added but not tested

**Review by mpidcock — APPROVED** [review]
Looks good + love to see the related cleanup and test improvements, just some minor comments

File: app/lib/pdf_filler/id40_pdf.rb [hunk 5]
```diff
@@ -43,6 +43,17 @@ def hash_for_pdf
...
+        'L12aSpouse' => @xml_document.at('SpouseOver65')&.text == "1" ? "Yes" : "Off",
+        'L12bYourself' => @xml_document.at('PrimeBlind')&.text == "1" ? "Yes" : "Off",
+        'L12bSpouse' => @xml_document.at('SpouseBlind')&.text == "1" ? "Yes" : "Off",
+        'L12cDependent' => @xml_document.at('ClaimedAsDependent')&.text == "1" ? "Yes" : "Off",
+        'TxCompL15' => 0,
+        'TxCompL16' => @xml_document.at('StandardDeduction')&.text,
+        'TxCompL17' => @xml_document.at('TaxableIncomeState')&.text,
+        'TxCompL19' => @xml_document.at('TaxableIncomeState')&.text,
+        'TxCompL20' => round_amount_to_nearest_integer(@xml_document.at('StateIncomeTax')&.text),
+        'L21' => round_amount_to_nearest_integer(@xml_document.at('StateIncomeTax')&.text),
```
[the remark this answers is not in the document]
Comment (jenny-heath) [reply app/lib/pdf_filler/id40_pdf.rb:56]: ah, so `.text` isn't converting it to a string, it's getting the contexts of the xml element, which is what's returned from `@xml_document.at('<node name>`)`. i'm not sure there is a way to get it out as a number, or at least i haven't seen one?

re: rounding, the xml accepts cents for this field (`AmountType`) but the pdf wants a whole number

[the remark this answers is not in the document]
Comment (mpidcock) [reply app/lib/pdf_filler/id40_pdf.rb:56]: that's the first I've heard of xml wanting/accepting cents but sometimes I guess I have to accept that things are just weird

**Pull request merged**